### ISSUE TITLE
Testing component with useParams hook returns undefined

### ISSUE THREAD

**Issue by lucksp** [testing-library/react-testing-library#654]
This has repeated for me in 2 different codebases & able to replicate in a simple code-sandbox too:

- `@testing-library/react` version:     `10.0.4` (see codesandbox below), and private repo for work using `9.5.0`, as well as in previous job codebase
- `react` version: 16.3, 16.8.3
- `node` version: 10.20 & 12 in previous job codebase
- `npm` (or `yarn`) version: 6.14

### Relevant code or config:
https://codesandbox.io/s/react-router-useparam-hook-test-s9jz1?file=/__tests__/Topic.test.js:792-991

### What you did:

I am trying to test components which are using the `useParams` react-router hook.  As the codebase is part of a production repository at work, we will need repeatable code utilities, so I used the recommended React-Testing-Library r[educe boilerplate](https://testing-library.com/docs/example-react-router#reducing-boilerplate) suggestion to write a wrapper with our Providers.  
When I run the tests, they seem OK, except for when I want to test a more complex component, which happens to use the `useParams` hook.  In this case the hook returns `undefined` each time so the component test fails.  However, if I test the higher up parents with the same route, it passes, but I don't want to always test higher up.

### What happened:

Test results: the test directly using the `useParam` hook returns `undefined`
```
I am route inside renderWithProviders Wrapper: /account/iamparam 
I am inside the Component and the param is:  iamparam  // Parent test passes here
I am route inside renderWithProviders Wrapper: /account/iamparam 
I am inside the Component and the param is:  undefined  // Child TEST FAILS HERE
```

After trying a modification to the `renderWithProviders` wrapper, to add an additional argument to accept a route param, this also did not pass:

```
function renderWithProviders(
  ui,
  {
    route = '/',
    params = routes.root,  // <== added new passed function param 
    history = createMemoryHistory({ initialEntries: [route] }),
  } = {},
) {
  return {
    ...render(
        <Router history={history}>
            <Route path={params}> // ==> tried to set path to "params" not "route" so the slug of /url/:accountId is properly set
             {ui}
            </Route>
        </Router>
    ),
    history,
  };
}

```

### Reproduction:
https://codesandbox.io/s/react-router-useparam-hook-test-s9jz1?file=/__tests__/Topic.test.js:792-991

### Problem description:

If the useParam hook works in some cases, but not others, it seems like the test library is not properly initializing something causing inconsistent results.

### Suggested solution:
It feels like something to do with the initial render cycle?

**Comment by lucksp** [comment]
I have seen another report of this here; https://stackoverflow.com/questions/59311255/how-to-create-unit-test-for-url-params-for-react-router-using-enzyme

**Comment by hd4ng** [comment]
@lucksp 
Try to set the path like the following:
```js
path='/account/:accountId'
```

You can read this: https://medium.com/@aarling/mocking-a-react-router-match-object-in-your-component-tests-fa95904dcc55
I had same issue, and the above article solve it.

**Comment by tachkoB** [comment]
I have a very similar issue, only with useRouteMatch. 

The component using the hook is not a direct child of the route, and when I try to test it, I get undefined for match. If I test the parents, I can see that the component is rendered, but testing the component in isolation, doesn't work.

What I did realize is, if I import useRouteMatch from "react-router" and not "react-router-dom", as I normally do, my tests pass, and I can log the path I have given to the component in the test. However, my app breaks and dynamic url's can no longer be built.

The opposite happens if Import it from "react-router-dom" my app works, the logs work, but tests get undefined.

**Comment by tachkoB** [comment]
so, it finally works for me.

what I had to do is in my tests 

import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';

and then:
  it('should render the component', () => {
    const history = createMemoryHistory();
    const route = '/whatever-the-route-is';
    history.push(route);

    const { getByText } = render(
      <Router history={history}>
        <ComponentToTest myProps={myProps} />
      </Router>,
    );
  });

so basically, you need to create the history context thta does the pushing. that way it is no longer undefined. enjoy!

**Comment by thril** [comment]
@tachkoB Your solution, along with @lucksp 's solution, did not work for me. This issue was closed, but I don't see a resolution anywhere. Can this be re-opened or more info provided? Thx.

**Comment by tachkoB** [comment]
> @tachkoB Your solution, along with @lucksp 's solution, did not work for me. This issue was closed, but I don't see a resolution anywhere. Can this be re-opened or more info provided? Thx.

you'd need to give more context.

have you tried
`    <MemoryRouter initialEntries={['blogs/1']}>
      <Route path='blogs/:blogId'>
        {children}
      </Route>
    </MemoryRouter>
`
from [stackoverflow](https://stackoverflow.com/questions/58117890/how-to-test-components-using-new-react-router-hooks)

**Comment by thril** [comment]
Yes, I tried the `MemoryRouter` with `initialEntries` as well, but the same results. I ended up using the `jest.mock` solution on that same SO page. While definitely not ideal, none of the other solutions worked.

**Comment by lucksp** [comment]
@thril  if you could provide more code snippets of your attempts that would help

**Comment by bryantee** [comment]
@hd4ng Thanks for the suggestion! That utility function from the article has been a great addition to our codebase.

**Comment by Tottzi** [comment]
> so, it finally works for me.
> 
> ```
> import { Router } from 'react-router-dom';
> import { createMemoryHistory } from 'history';
> 
>   it('should render the component', () => {
>     const history = createMemoryHistory();
>     const route = '/whatever-the-route-is';
>     history.push(route);
> 
> 
>     const { getByText } = render(
>       <Router history={history}>
>         <ComponentToTest myProps={myProps} />
>       </Router>,
>     );
>   });
> ```
> 
> so basically, you need to create the history context that does the pushing. that way it is no longer undefined. enjoy!

But it won't work if you're using already BrowserRouter for your app. :(
